### PR TITLE
Resolve assembly location for SingleFile

### LINKED ISSUE #1685: Support for SingleFile && SelfContained apps
Currently, it seems like the true single file deployment (supported in .NET 5 onwards) is not working properly with BenchmarkDotNet.

### Repro

```sh
# on Unbuntu 20.04 with latest .NET 5.0 (5.0.4) in PATH
$ dotnet new console -n MyBenchmarkApp
$ cd MyBenchMarkApp
$ dotnet add package BenchmarkDotNet

# write some InProcess benchmarks
$ cat > Program.cs <<EOF
using System;
using System.Threading;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

[InProcess]
public class Program
{
    [Benchmark]
    public void Slow() => Thread.Sleep(100);

    [Benchmark]
    public void Fast() => Thread.Sleep(25);

    static void Main() => BenchmarkRunner.Run<Program>();
}
EOF

# deploy as a singlefile ELF executable (to a dir called ./publish)
$ dotnet publish -c Release -r linux-x64 --self-contained true -p:PublishSingleFile=true -o publish

# first test the (shared) executable which gets generated as part of the build under `./bin`
$ bin/Release/net5.0/linux-x64/MyBenchmarkApp
# ✅ works!
# and size of this shared executable
$ du -h bin/Release/net5.0/linux-x64/MyBenchmarkApp
11M     bin/Release/net5.0/linux-x64/MyBenchmarkApp

# then try the published app
$ publish/MyBenchmarkApp
# 🛑 fails!
// Validating benchmarks:
// ***** BenchmarkRunner: Start   *****
// ***** Found 2 benchmark(s) in total *****
// ***** Building 1 exe(s) in Parallel: Start   *****
Failed to generate partition: System.ArgumentNullException: Value cannot be null. (Parameter 'path1')
   at System.IO.Path.Combine(String path1, String path2)
   at BenchmarkDotNet.Toolchains.InProcess.Emit.InProcessEmitGenerator.GetArtifactsPaths(BuildPartition buildPartition, String rootArtifactsFolderPath)
   at BenchmarkDotNet.Toolchains.InProcess.Emit.InProcessEmitGenerator.GenerateProject(BuildPartition buildPartition, ILogger logger, String rootArtifactsFolderPath)
// ***** Done, took 00:00:00 (0.11 sec)   *****
// Found 2 benchmarks:
//   Program.Slow: InProcess(Toolchain=InProcessEmitToolchain)
//   Program.Fast: InProcess(Toolchain=InProcessEmitToolchain)

// Build Error: Exception!
Message: Object reference not set to an instance of an object.,
Stack trace:
   at BenchmarkDotNet.Toolchains.Results.BuildResult.Failure(GenerateResult generateResult, Exception exception)
   at BenchmarkDotNet.Running.BenchmarkRunnerClean.Build(BuildPartition buildPartition, String rootArtifactsFolderPath, ILogger buildLogger)

// Build Error: Exception!
Message: Object reference not set to an instance of an object.,
Stack trace:
   at BenchmarkDotNet.Toolchains.Results.BuildResult.Failure(GenerateResult generateResult, Exception exception)
   at BenchmarkDotNet.Running.BenchmarkRunnerClean.Build(BuildPartition buildPartition, String rootArtifactsFolderPath, ILogger buildLogger)

// ***** BenchmarkRunner: Finish  *****
...

# and size of this single executable
$ du -h publish/MyBenchmarkApp
77M     publish/MyBenchmarkApp
```

another strange thing is the `./publish` directory has more files than expected (in case of a console app without BenchmarkDotNet reference):

```sh
$ ls -R publish
publish:
amd64  MyBenchmarkApp MyBenchmarkApp.pdb  x86

publish/amd64: // is it necessary on Linux?
KernelTraceControl.dll  msdia140.dll

publish/x86: // is it necessary on Linux?
KernelTraceControl.Win61.dll  KernelTraceControl.dll  msdia140.dll
```

### PR DESCRIPTION
**Pull request by am11**
Fixes #1685.

### CODE REVIEW

**Comment by am11** [comment]
Same tests failed with master branch (with just two whitespaces in the readme)
change: https://github.com/dotnet/BenchmarkDotNet/compare/master...e9d1e04
checks: https://github.com/dotnet/BenchmarkDotNet/pull/1686/checks?check_run_id=2266431063
(reverting the test commit)

**Comment by adamsitnik** [comment]
@am11 I've reported the CoreRT error: https://github.com/dotnet/runtimelab/issues/927

**Comment by adamsitnik** [comment]
the tests should be passing now, I am going to close and reopen the PR to kick the CI legs

File: src/BenchmarkDotNet/Running/BuildPartition.cs [hunk 1]
```diff
@@ -60,5 +62,11 @@ public BuildPartition(BenchmarkBuildInfo[] benchmarks, IResolver resolver)
         public bool IsCustomBuildConfiguration => BuildConfiguration != InfrastructureMode.ReleaseConfigurationName;
 
         public override string ToString() => RepresentativeBenchmarkCase.Job.DisplayInfo;
+
+        private static string GetResolvedAssemblyLocation(Assembly assembly) =>
+            // in case of SingleFile, location.Length returns 0, so we use GetName() and
+            // manually construct the path.
+            assembly.Location.Length == 0 ?
+                Path.Combine(AppContext.BaseDirectory, assembly.GetName().Name) : assembly.Location;
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Running/BuildPartition.cs:70]: just to trigger the CI

```suggestion
            assembly.Location.Length == 0 ? Path.Combine(AppContext.BaseDirectory, assembly.GetName().Name) : assembly.Location;
```

File: src/BenchmarkDotNet/Toolchains/CsProj/CsProjGenerator.cs [hunk 2]
```diff
@@ -35,7 +35,7 @@ public CsProjGenerator(string targetFrameworkMoniker, string cliPath, string pac
 
         protected override string GetBuildArtifactsDirectoryPath(BuildPartition buildPartition, string programName)
         {
-            string assemblyLocation = buildPartition.AssemblyLocation;
+            string assemblyLocation = buildPartition.RepresentativeBenchmarkCase.Descriptor.Type.Assembly.Location;
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Toolchains/CsProj/CsProjGenerator.cs:38]: why are you not using `buildPartition.AssemblyLocation` here?

Comment (am11) [reply src/BenchmarkDotNet/Toolchains/CsProj/CsProjGenerator.cs:38]: This is a separate mechanism for “assembly loaded from stream”. I didn’t wanted to change/refactor it too much, as here we expect Location to return null for LoadFromStream case and test fails if Location is non-null.

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @am11 !

**Pull request merged**